polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -44,6 +45,50 @@ using namespace mlir::polygeist;
 
 namespace {
 
+//===----------------------------------------------------------------------===//
+// WorkGroupSize
+//===----------------------------------------------------------------------===//
+
+/// Class to represent work group size of a kernel.
+class WorkGroupSize {
+public:
+  using ElemTy = std::variant<Value, unsigned>;
+  WorkGroupSize(Value ndItem, const sycl::ReqdWorkGroupSize &reqdWorkGroupSize,
+                OpBuilder builder) {
+    auto ndItemTy = cast<sycl::NdItemType>(
+        cast<MemRefType>(ndItem.getType()).getElementType());
+    for (unsigned dim = 0; dim < ndItemTy.getDimension(); ++dim) {
+      /// Use constants provided by reqd_work_group_size if given (non-empty).
+      if (!reqdWorkGroupSize.empty())
+        wgSizes.push_back(reqdWorkGroupSize[dim]);
+      else
+        wgSizes.push_back(builder.create<arith::IndexCastOp>(
+            builder.getUnknownLoc(), builder.getIndexType(),
+            getLocalRange(ndItem, dim, builder)));
+    }
+  }
+
+  /// Return true if the element holds the alternative T.
+  template <typename T> bool is() const {
```

Comment:
T should be one of `Value` or `unsigned`.  Can you use: 
```
template <typename T,
          typename =
              std::enable_if_t<llvm::is_one_of<T, Value, unsigned>::value>>
```

or a static_assert to limit the value T may have.